### PR TITLE
[WIP] Allow collapsing vars in async context

### LINKED ISSUE #528: pure_funcs calls are not inlined
**Bug report or Feature request?**

Feature request.

**Version (complete output of `terser -V` or specific git commit)**

4.3.4

**Complete CLI command or `minify()` options used**

```bash
npx terser temp.js -c pure_funcs=["Uint8Array"],passes=2
```

**`terser` input**

```js
(function(u) {
  console.log(u);
})([1, 2, 3]);

new Uint8Array([1, 2, 3]);

(function foo(u) {
  console.log(u);
})(new Uint8Array([1, 2, 3]));
```

**`terser` output or error**

```js
console.log([1,2,3]),function(u){console.log(u)}(new Uint8Array([1,2,3]));
```

**Expected result**

```js
console.log([1,2,3]),console.log(new Uint8Array([1,2,3]));
```

---

As you can see, `pure_funcs` correctly works for DCE, but for some reason it's not recognised as a constant that can be inlined in other functions.

### PR DESCRIPTION
**Pull request by RReverser**
Fixes #528.

### CODE REVIEW

**Comment by kzc** [comment]
Please gate this feature with a new compress option "unsafe_await" as discussed in #528. When it's tested in the wild on real life applications we might consider dropping the option if demonstrated to be safe.

I'd prefer that `doSomething` be changed to `console.log`, and pass a different array literal to each IIFE to make the test more robust. It's useful to have tests that are easily run from the command line for verification rather than reason about them and make a mistake. Once the test runner supports async operation in the future then an `expect_stdout` clause could easily be added to these tests, which would allow further regression testing with a variety of minify options in `test/ufuzz.json`.

**Comment by kzc** [comment]
The unrelated regression in master was caused by https://github.com/terser/terser/commit/942a9ef73a36c2a9f260be739be36e19633e732f

**Comment by RReverser** [comment]
> Please gate this feature with a new compress option "unsafe_await" as discussed in #528.

Hmm, I thought `unsafe_await` related only to inlining? Anything that doesn't have side-effects, including vars collapsing, is safe to do in any context (that's the whole point of async-await after all - to behave like synchronous syntax). So applying `unsafe_await` makes perfect sense for #529, but seems a bit overconservative for this transform.

**Comment by RReverser** [comment]
> I'd prefer that doSomething be changed to console.log, and pass a different array literal to each IIFE to make the test more robust. It's useful to have tests that are easily run from the command line for verification rather than reason about them and make a mistake.

Makes sense, I'll update tests later.

**Comment by kzc** [comment]
Unfortunately this test fails using this PR. Please mark PR as [WIP].

```
$ cat abug.js
var x = 100;

function side_effect() {
    return console.log(++x), x;
}

function print(x) {
    console.log(x);
}

(async function(e) {
    var f = side_effect();
    await print(f);
    print(e);
})([2]);

(async function(e) {
    var f = side_effect();
    await 42;
    print(f);
    print(e);
})([3]);
```
```
$ cat abug.js | node-v12.13.0 
101
101
102
[ 2 ]
102
[ 3 ]
```
```
$ cat abug.js | terser -bc | node-v12.13.0 
101
101
[ 2 ]
102
102
[ 3 ]
```

**Comment by kzc** [comment]
> Anything that doesn't have side-effects, including vars collapsing, is safe to do in any context (that's the whole point of async-await after all - to behave like synchronous syntax).

async/await is just sugar. It appears to be synchronous code, but it is not.

`reduce_vars` deals with constant propagation, whereas `collapse_vars` propagates non-constant variable payloads.

**Comment by RReverser** [comment]
> Please mark PR as [WIP].

Done. I just implemented the patch the way you suggested - feel free to make edits. I'm not very familiar with terser codebase and it's a bit hard to follow tbh :(

> async/await is just sugar. It appears to be synchronous code, but it is not.

I know how async/await works. What I'm saying is that for purposes of code minification and, in particular, constant and pure expression propagation, it can be treated in the same way as synchronous code.

**Comment by kzc** [comment]
As previously mentioned, the code in question does not deal with constant propagation. The readme states:

- `collapse_vars` (default: `true`) -- Collapse single-use non-constant variables,
  side effects permitting.

The [patch](https://github.com/terser/terser/issues/528#issuecomment-562876860) just identified why the value in your specific example wasn't being simplified into the await expression. I never suggested it's a final solution - it's just a starting point. I'll leave this with you. I could implement it easily enough, but better to get the people actually interested in the feature to do it. Otherwise the project will stagnate. A half dozen people have learned the code well enough to make significant contributions. It's not insurmountable. No more or less difficult than other optimizing compilers.

**Comment by RReverser** [comment]
> A half dozen people have learned the code well enough to make significant contributions.

Sure, but, like you, I have other OSS projects to maintain and take care of, as well as a daily job :) Taking on and digging into a new OSS project can be fun, but I wouldn't underestimate amount of time and effort it requires, and certainly shouldn't be motivated by "I just miss this one specific feature".

With that in mind, I completely understand if you don't have time and/or interest to work on this either, in which case feel free to either close it or leave the original issue open for other potential contributors.

**Comment by kzc** [comment]
Fair enough. Keep the issues and PR open for anyone interested. Who knows - some other optimization might spark your interest and you might revisit it.

**Comment by fabiosantoscode** [comment]
@RReverser this is a pretty good idea! Thanks for sending this PR.

However it's going to be a tad complex to work this out, I feel. Maybe it's easier to collapse variables between awaits, but not across awaits.

**Comment by RReverser** [comment]
@fabiosantoscode I'm not sure I understand the difference.

**Comment by fabiosantoscode** [comment]
@RReverser

By between awaits I mean this situation:

```
await something()
const a = 42
doSomething(a)
await somethingElse()
```

By "across awaits" (probably not the best term!) I mean something like:

```
const a = 42
await somethingElse()
doSomething(a)
```

Where `await` is in the way.

So I probably should've:

 - Said "when await is in the way" instead of "across awaits"
 - Replied to you earlier than May! Sorry about that.


This principle of checking if something is in the way before inlining also applies to:

 - yield
 - license comments marked as important (to avoid incompatibly licensed code being inlined together)

**Comment by RReverser** [comment]
I see. But I'm not sure I understand the distinction / why we don't want to do this across `await` as well. `await` can't affect local variables any more than a regular function call can, so it should be equally safe and easy to inline any vars whether they have `await` in between or not...

**Comment by fabiosantoscode** [comment]
`await` doesn't affect local (or in-scope) variables itself, but it lets other things run which might.

Take this function:

```
let fooSource = []         
                           
async function x() {       
    const foo = fooSource  
    await null             
    return foo[0]          
}                          
                           
const promise = x()                        
                           
fooSource[0] = 'changed!'  

console.log('result:', promise)
```

Take the `await` away and the result is "changed".

Terser assumes that a function runs uninterrupted while inlining things, and `await` breaks this assumption. I think that's why [this failed](https://github.com/terser/terser/pull/530#issuecomment-564666000).

**Comment by RReverser** [comment]
Right, I see what you mean. But presumably that's what already happens when calling other functions - could `await` be treated as if it was a call to some function in outer scope?

**Comment by fabiosantoscode** [comment]
We actually could just say it "has side effects" and call it a day, but it's one of those bandaids that can be troublesome to maintain, understand and debug in the long run.

**Comment by RReverser** [comment]
That actually sounds pretty good and would already cover lots of cases probably.
